probe-rs/src/architecture/arm/sequences/nxp_armv8m.rs

```diff
@@ -1,3 +1,358 @@
+            interface.write_raw_dp_register(dp, DP_CTRL_STAT, 0x50000F00)?;
+
+            // Clear WDATAERR, STICKYORUN, STICKYCMP, and STICKYERR bits of CTRL/STAT Register by write to ABORT register
+            interface.write_raw_dp_register(dp, SW_DP_ABORT, 0x0000001E)?;
+
+            let ap = ApAddress { dp, ap: 0 };
+            let mem_ap = MemoryAp::new(ap);
+            self.enable_debug_mailbox(interface, dp, mem_ap)?;
+        }
+
+        tracing::trace!("MIMXRT5xxS debug port start was successful");
+
+        Ok(())
+    }
+
+    fn reset_system(
+        &self,
+        probe: &mut dyn ArmProbe,
+        core_type: probe_rs_target::CoreType,
+        _debug_base: Option<u64>,
+    ) -> Result<(), ArmError> {
+        self.check_core_type(core_type)?;
+
+        tracing::trace!("MIMXRT5xxS reset system");
+
+        // Halt the core
+        let mut dhcsr = Dhcsr(0);
+        dhcsr.set_c_halt(true);
+        dhcsr.set_c_debugen(true);
+        dhcsr.enable_write();
+        probe.write_word_32(Dhcsr::get_mmio_address(), dhcsr.into())?;
+        probe.flush()?;
+
+        // Clear VECTOR CATCH and set TRCENA
+        let mut demcr: Demcr = probe.read_word_32(Demcr::get_mmio_address())?.into();
+        demcr.set_trcena(true);
+        demcr.set_vc_corereset(false);
+        probe.write_word_32(Demcr::get_mmio_address(), demcr.into())?;
+        probe.flush()?;
+
+        // Reset the flash peripheral on FlexSPI0, if any.
+        self.reset_flash(probe)?;
+
+        // Set watch point at SYSTEM_STICK_CALIB access
+        probe.write_word_32(Self::DWT_COMP0, Self::SYSTEM_STICK_CALIB_ADDR)?;
+        probe.write_word_32(Self::DWT_FUNCTION0, 0x00000814)?;
+        probe.flush()?;
+        tracing::trace!("set data watchpoint for MIMXRT5xxS reset");
+
+        // Execute SYSRESETREQ via AIRCR
+        let mut aircr = Aircr(0);
+        aircr.set_sysresetreq(true);
+        aircr.vectkey();
+        // (we need to ignore errors here because the reset will make this
+        // operation seem to have failed.)
+        probe
+            .write_word_32(Aircr::get_mmio_address(), aircr.into())
+            .ok();
+        probe.flush().ok();
+
+        tracing::trace!("MIMXRT5xxS reset system was successful; waiting for halt after reset");
+
+        self.wait_for_stop_after_reset(probe)
+    }
+
+    fn reset_hardware_deassert(&self, memory: &mut dyn ArmProbe) -> Result<(), ArmError> {
+        tracing::trace!("MIMXRT5xxS reset hardware deassert");
+        let n_reset = Pins(0x80).0 as u32;
+
+        let can_read_pins = memory.swj_pins(0, n_reset, 0)? != 0xffff_ffff;
+
+        thread::sleep(Duration::from_millis(50));
+
+        let mut assert_n_reset = || memory.swj_pins(n_reset, n_reset, 0);
+
+        if can_read_pins {
+            let start = Instant::now();
+            let timeout_occured = || start.elapsed() > Duration::from_secs(1);
+
+            while assert_n_reset()? & n_reset == 0 || !timeout_occured() {
```

Comment:
What was the original intention here? `while !timeout_occurred()` is a bit of a weird loop, it looks like it'll wait for a second in all cases.

Reply:
Agreed that this seems strange. I'm afraid I don't remember what (if anything) I was thinking here. Perhaps this was a mistake and should've been `&&` instead of `||`.